File: client/java/Unit.java
import com.sun.jna.Pointer;

///Generic Unit
abstract class Unit extends Mappable 
{
  public Unit(Pointer p)
  {
    super(p);
  }
  abstract boolean validify();
    
    //commands
    
  ///Move the unit to the designated X and Y coordinates if possible
  int move(int x, int y)
  {
    validify();
    return Client.INSTANCE.unitMove(ptr, x, y);
  }
  ///Allows a unit to display a message to the screen.
  int talk(String message)
  {
    validify();
    return Client.INSTANCE.unitTalk(ptr, message);
  }
  ///Attempt to attack the input target if possible
  int attack(Unit Target)
  {
    validify();
    Target.validify();
    return Client.INSTANCE.unitAttack(ptr, Target.ptr);
  }
    
    //getters
    
  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.unitGetId(ptr);
  }
  ///The X position of this object.  X is horizontal, with 0,0 as the top left corner
  public int getX()
  {
    validify();
    return Client.INSTANCE.unitGetX(ptr);
  }
  ///The Y position of this object.  Y is vertical, with 0,0 as the top left corner
  public int getY()
  {
    validify();
    return Client.INSTANCE.unitGetY(ptr);
  }
  ///Represents the owner of the unit.
  public int getOwner()
  {
    validify();
    return Client.INSTANCE.unitGetOwner(ptr);
  }
  ///Current ealth of the unit
  public int getHealth()
  {
    validify();
    return Client.INSTANCE.unitGetHealth(ptr);
  }
  ///Attacking strength of the unit (Each point of strength deals 1 health of damage)
  public int getStrength()
  {
    validify();
    return Client.INSTANCE.unitGetStrength(ptr);
  }
  ///Displays the remaining moves for this unit this turn
  public int getMovesLeft()
  {
    validify();
    return Client.INSTANCE.unitGetMovesLeft(ptr);
  }
  ///Displays the remaining attacks for this unit this turn
  public int getAttacksLeft()
  {
    validify();
    return Client.INSTANCE.unitGetAttacksLeft(ptr);
  }
  ///Amount of gold carried by the unit.
  public int getGold()
  {
    validify();
    return Client.INSTANCE.unitGetGold(ptr);
  }

}

File: client/c/Unit.cpp
// -*-c++-*-

#include "Unit.h"
#include "game.h"

#include "Unit.h"

Unit::Unit(_Unit* pointer)
{
    ptr = (void*) pointer;
}

int Unit::id()
{
  return ((_Unit*)ptr)->id;
}

int Unit::x()
{
  return ((_Unit*)ptr)->x;
}

int Unit::y()
{
  return ((_Unit*)ptr)->y;
}

int Unit::owner()
{
  return ((_Unit*)ptr)->owner;
}

int Unit::health()
{
  return ((_Unit*)ptr)->health;
}

int Unit::strength()
{
  return ((_Unit*)ptr)->strength;
}

int Unit::movesLeft()
{
  return ((_Unit*)ptr)->movesLeft;
}

int Unit::attacksLeft()
{
  return ((_Unit*)ptr)->attacksLeft;
}

int Unit::gold()
{
  return ((_Unit*)ptr)->gold;
}


int Unit::move(int x, int y)
{
  return unitMove( (_Unit*)ptr, x, y);
}

int Unit::talk(char* message)
{
  return unitTalk( (_Unit*)ptr, message);
}

int Unit::attack(Unit& Target)
{
  return unitAttack( (_Unit*)ptr, (_Unit*) Target.ptr);
}



std::ostream& operator<<(std::ostream& stream,Unit ob)
{
  stream << "id: " << ((_Unit*)ob.ptr)->id  <<'\n';
  stream << "x: " << ((_Unit*)ob.ptr)->x  <<'\n';
  stream << "y: " << ((_Unit*)ob.ptr)->y  <<'\n';
  stream << "owner: " << ((_Unit*)ob.ptr)->owner  <<'\n';
  stream << "health: " << ((_Unit*)ob.ptr)->health  <<'\n';
  stream << "strength: " << ((_Unit*)ob.ptr)->strength  <<'\n';
  stream << "movesLeft: " << ((_Unit*)ob.ptr)->movesLeft  <<'\n';
  stream << "attacksLeft: " << ((_Unit*)ob.ptr)->attacksLeft  <<'\n';
  stream << "gold: " << ((_Unit*)ob.ptr)->gold  <<'\n';
  return stream;
}

